### PR TITLE
Autocomplete dropdown ( and much, much more....)

### PR DESCRIPTION
**Pull request by jacobfelknor**
This is a big one, bear with me

Some improvements:
- Added department model. General database normalization reasons
- Department field is now an autocomplete field (no dropdown functionality for now) in the search forms. Can now search both on a major code (APPM, CSCI, etc) or the name (Applied Math, Computer Science, etc). This field was updated in the user signup/edit form as well.
- Tables now scroll. Will dynamically grab screen size and set that as max height for div. If results are longer, scroll the table. This accomplishes the sticky-top functionality with more stability. (sticky top messed with the form autocomplete. The autocomplete choices would display behind the form, which is not acceptable.)
- Searches are now stored in HTML5's native sessionStorage interface. This means that searches persist in the fields when you return to them, and will only reset when the reset button is pressed or you start a new session (new browser tab)
- As mentioned, reset button to clear all search params currently entered (also clears them from the sessionStorage)
- Rewrote quite a bit of FCQ/Professor server side processing to be more consistant with the class search functions
- Bug fix where you could not view a class if you weren't logged in (this bug started when the schedule object was added)
- Added site-wide background image for visual appeal
- Wrote a few more test cases
- More I'm sure that I just can't remember right now

I've also made some proposed changes for the way classes/schedules/teachers/fcqs relate to one another in the database (want to normalize as much as possible) but have not implemented these yet because I want to discuss before major changes are made. You can see these proposed changes in the models.py files of the classes/fcq apps. Also, this pull request is already massive so I didn't want to complicate it further. 

Let me know if you have questions/comments/suggestions!!!

### CODE REVIEW

File: classes/models.py [hunk 1]
```diff
@@ -3,14 +3,49 @@
...
+        end_time
+        instructor (foreign key to instructor)
+        days
+        session
+        semester (Spring/Fall/Summer <year>)
+        max_enrollment
+        building_room
+        campus
+        schedule (Many to Many Field)
+"""
```
Comment (PaperXLV) [inline classes/models.py]: As I understand, Classes are more persistent (i.e. we use the same class that holds CSCI 2270 for multiple semesters), and then sections hold the actual data for the semester? I could see that being a good change

Also one small thought on the class search functionality, I think the lowest value section number should show up on the search page, that way the user is taken first to the 'most important' section. (I find I'm occasionally brought to recitations first and I'm most interested in seeing the lecture times)

File: fcq/models.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+Foreign key back to the professor which coresponds
+with the FCQ.
+
+Change department and course to Foreign Keys
+to their respective models
+(this is dependent on the database changes
+happening for the class models, see notes there.
+If that doesn't happen, leave these as Chars)
+
+"""
```
Comment (PaperXLV) [inline fcq/models.py]: I agree with the first change, that way if we remove or add fcqs the averages for the professors change as we would then calculate it.

I'm curious what the use case for having a foreign key back to the professor is, I would say that change makes sense though.

And I agree with the class changes, so dept and course should be foreign keys. (notably not section, I don't think that would be feasible or usable)

File: fcq/views.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        )  # this feels sort of hackish...
+        number_query = Q(courseList__contains=[number])
+        query &= number_query
+    teachers = Teacher.objects.filter(query).order_by("lastName")
+    response = TeacherSerializer(teachers, many=True)
+    return JsonResponse(response.data, safe=False)
+
+
+class TeacherView(DetailView):
+    model = Teacher
```
Comment (PaperXLV) [inline fcq/views.py]: not quite sure where the line of code is, I think the number picker on course number is a little goofy. Users would either know an exact class to put in there, or they would leave it blank. I think it should just be text

File: classes/models.py [hunk 4]
```diff
@@ -3,14 +3,49 @@
...
+class Department(models.Model):
+    name = models.CharField(max_length=200)
+    code = models.CharField(max_length=4)
 
 
 class Class(models.Model):
 
     # Database
-    department = models.CharField(max_length=4)
+    # department = models.CharField(max_length=4)
```
[the remark this answers is not in the document]
Comment (jacobfelknor) [reply classes/models.py]: done

File: fcq/urls.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+from django.urls import path
+
+from . import views
+
+app_name = "fcq"
+
+urlpatterns = [
+    path("fcq_search/", views.fcq_search, name="fcq_search"),
+    path("fcq_search_ajax/", views.fcq_search_ajax, name="fcq_search_ajax"),
+    # path("fcq_display/", views.fcq_display, name="fcq_display"), # unused for now
```
[the remark this answers is not in the document]
Comment (jacobfelknor) [reply fcq/urls.py]: yes. done

**Pull request merged**